pkg/volume/csi/nodeinfomanager/nodeinfomanager.go

```diff
@@ -74,11 +80,108 @@ type Interface interface {
 func NewNodeInfoManager(
 	nodeName types.NodeName,
-	volumeHost volume.VolumeHost) Interface {
+	volumeHost volume.VolumeHost,
+	migratedDrivers map[string](func() bool)) Interface {
 	return &nodeInfoManager{
-		nodeName:   nodeName,
-		volumeHost: volumeHost,
+		nodeName:        nodeName,
+		volumeHost:      volumeHost,
+		migratedDrivers: migratedDrivers,
+	}
+}
+
+func (nim *nodeInfoManager) InitializeMigratedDrivers() error {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.CSINodeInfo) {
+		return nil
+	}
+
+	csiKubeClient := nim.volumeHost.GetCSIClient()
+	if csiKubeClient == nil {
+		return fmt.Errorf("error getting CSI client")
+	}
+
+	retryErr := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		nodeInfo, err := csiKubeClient.CsiV1alpha1().CSINodeInfos().Get(string(nim.nodeName), metav1.GetOptions{})
+		if nodeInfo == nil || errors.IsNotFound(err) {
+			nodeInfo, err = nim.CreateCSINodeInfo()
+		}
+		if err != nil {
+			return err // do not wrap error
+		}
+		return nim.initializeMigratedDrivers(nodeInfo)
+	})
+	if retryErr != nil {
+		return fmt.Errorf("CSINodeInfo update failed: %v", retryErr)
+	}
+	return nil
+}
+
+func (nim *nodeInfoManager) initializeMigratedDrivers(nodeInfo *csiv1alpha1.CSINodeInfo) error {
+	csiKubeClient := nim.volumeHost.GetCSIClient()
+	if csiKubeClient == nil {
+		return fmt.Errorf("error getting CSI client")
+	}
+
+	existingDrivers := sets.String{}
+
+	// Do existing updates
+	// Clone driver list, omitting the driver that matches the given driverName
+	newDriverSpecs := []csiv1alpha1.CSIDriverInfoSpec{}
+	for _, driverInfoSpec := range nodeInfo.Spec.Drivers {
+		// No need to change any specs
+		newDriverSpecs = append(newDriverSpecs, driverInfoSpec)
+		existingDrivers.Insert(driverInfoSpec.Name)
+	}
+
+	newDriverStatuses := []csiv1alpha1.CSIDriverInfoStatus{}
+	for _, driverInfoStatus := range nodeInfo.Status.Drivers {
+		if _, ok := nim.migratedDrivers[driverInfoStatus.Name]; ok {
+			// If driver is known to migration logic and already exists in Spec
+			// just update volume plugin mechanism
+			driverInfoStatus.VolumePluginMechanism = nim.getVolumePluginMechanism(driverInfoStatus.Name)
+		}
+		newDriverStatuses = append(newDriverStatuses, driverInfoStatus)
 	}
+
+	for driver, _ := range nim.migratedDrivers {
+		if existingDrivers.Has(driver) {
+			// Skip adding new driver because already updated previously
+			continue
+		}
+
+		driverSpec := csiv1alpha1.CSIDriverInfoSpec{
+			Name:         driver,
+			NodeID:       "",
+			TopologyKeys: []string{},
+		}
+
+		volumePluginMechanism := nim.getVolumePluginMechanism(driver)
+		driverStatus := csiv1alpha1.CSIDriverInfoStatus{
+			Name: driver,
+			// We are just adding this driver object it's not installed
+			Available:             false,
+			VolumePluginMechanism: volumePluginMechanism,
+		}
+
+		newDriverSpecs = append(newDriverSpecs, driverSpec)
+		newDriverStatuses = append(newDriverStatuses, driverStatus)
```

Comment:
As per offline discussion today:

My plan is to continue to work on it and get it in with some form of the current "updating spec and status together" code smell. With a big TODO/Issue linked so that it will all get addressed in a Q1 refactor which will introduce the reconciler concept to all the pieces of the nodeinfomanager. However, I will update the code to make two calls sequentially (one to the main resource for spec, one to the subresource for status) so I can shape the API now then split the code internally into reconcile loops later